src/utils/allocators.zig

```diff
@@ -178,120 +192,154 @@ pub const DiskMemoryAllocator = struct {
-};
 
-test "recycle allocator" {
-    const backing_allocator = std.testing.allocator;
-    var allocator = try RecycleFBA(.{}).init(backing_allocator, 1024);
-    defer allocator.deinit();
+    /// unmaps the memory (file still exists and is removed on deinit())
+    fn free(ctx: *anyopaque, buf: []u8, log2_align: u8, return_address: usize) void {
+        _ = return_address;
+        const self: *Self = @ptrCast(@alignCast(ctx));
 
-    // alloc a slice of 100 bytes
-    const bytes = try allocator.allocator().alloc(u8, 100);
-    const ptr = bytes.ptr;
-    // free the slice
-    allocator.allocator().free(bytes);
+        const alignment = @as(usize, 1) << @intCast(log2_align);
+        std.debug.assert(alignment <= std.mem.page_size); // the allocator interface shouldn't allow this (aside from the *Raw methods).
 
-    // realloc should be the same (ie, recycled data)
-    const bytes2 = try allocator.allocator().alloc(u8, 100);
-    try std.testing.expectEqual(ptr, bytes2.ptr);
-    allocator.allocator().free(bytes2);
+        const buf_ptr: [*]align(std.mem.page_size) u8 = @alignCast(buf.ptr);
+        const aligned_size = fullMmapSize(buf.len);
 
-    // same result with smaller slice
-    const bytes3 = try allocator.allocator().alloc(u8, 50);
-    try std.testing.expectEqual(ptr, bytes3.ptr);
-    allocator.allocator().free(bytes3);
+        const metadata: Metadata = @bitCast(buf_ptr[buf.len..][0..@sizeOf(Metadata)].*);
+        const file_name_bounded = fileNameBounded(metadata.file_index);
+        const file_name = file_name_bounded.constSlice();
 
-    // diff result with larger slice
-    const bytes4 = try allocator.allocator().alloc(u8, 200);
-    try std.testing.expect(ptr != bytes4.ptr);
-    allocator.allocator().free(bytes4);
-}
+        std.posix.munmap(buf_ptr[0..aligned_size]);
+        self.dir.deleteFile(file_name) catch |err| {
+            self.logFailure(err, file_name);
+        };
+    }
+
+    const Metadata = extern struct {
+        file_index: usize,
+    };
+
+    /// Returns the aligned size with enough space for `size` and `Metadata` at the end.
+    inline fn fullMmapSize(size: usize) usize {
+        return std.mem.alignForward(usize, size + @sizeOf(Metadata), std.mem.page_size);
+    }
+
+    fn logFailure(self: Self, err: anyerror, file_name: []const u8) void {
+        self.logger.errf("Disk Memory Allocator error: {s}, filepath: {s}", .{
+            @errorName(err), sig.utils.fmt.tryRealPath(self.dir, file_name),
+        });
+    }
+
+    const file_name_max_len = sig.utils.fmt.boundedLenValue("bin_{d}", .{std.math.maxInt(usize)});
+    inline fn fileNameBounded(file_index: usize) std.BoundedArray(u8, file_name_max_len) {
+        return sig.utils.fmt.boundedFmt("bin_{d}", .{file_index});
+    }
+};
 
 test "disk allocator on hashmaps" {
-    var allocator = DiskMemoryAllocator.init(sig.TEST_DATA_DIR ++ "tmp");
-    defer allocator.deinit(null);
+    var tmp_dir_root = std.testing.tmpDir(.{});
```

Comment:
even though the probability of two random dir being the same name is close to zero, would pref to have it a static string value -- see how we do blockstore tests

Reply:
What's the advantage? Static dirs have a bigger weakness imo: programmers can accidentally cause them to race by doing something as simple as copy-pasting code.

Also, to be clear, not only are the chances of two dirs having the same randomly having the same name essentially zero, but the temporary directory tree is deleted on cleanup, meaning the only time two tests could race is if they are running in parallel, in which case the likelihood of the crypto rng generating the same sequence twice is practically indistinguishable from zero.

Reply:
ah thats a good point - static name dirs do have a chance of overlapping if the programmer makes a mistake - random names makes more sense in that case

@dnut wdyt about changing the blockstore test code to use tmpDirs too?

Reply:
I tried to address that issue, and I've been having some more ideas to reduce the risk even further. But I can't think of a totally foolproof solution for that problem, unless you ensure a unique path is generated on every call, which tmpdir is good for.

My original goal was to leave around the state from the test in a place that's easy to find. That way I could inspect it if something goes wrong in the test. But actually, it's pretty rare that I need to do that. Usually I just re-run the entire test and debug it. Also, if I need to ad hoc inspect the state for failed tests, it's not that hard to change the test to use a known path. So I'm fine with using tmp dirs if that's what you guys prefer.

Reply:
yeah would pref that, whenever we can, pretty low priority rn - created a ticket : https://github.com/orgs/Syndica/projects/2/views/1?pane=issue&itemId=78811926